pkg/controller/common/reconciler/secret.go

```diff
@@ -41,3 +51,120 @@ func ReconcileSecret(c k8s.Client, expected corev1.Secret, owner metav1.Object)
 	}
 	return reconciled, nil
 }
+
+// ReconcileSecretNoOwnerRef should be called to reconcile a Secret for which we explicitly don't want
+// an owner reference to be set, and want existing ownerRefs from previous operator versions to be removed,
+// because of this k8s bug: https://github.com/kubernetes/kubernetes/issues/65200 (fixed in k8s 1.20).
+//
+// It makes sense to use this function for secrets which are likely to be manually
+// copied into other namespaces by the end user.
+// Because of the k8s bug mentioned above, the ownerRef could trigger a racy garbage collection
+// that deletes all children resources, potentially resulting in data loss.
+// See https://github.com/elastic/cloud-on-k8s/issues/3986 for more details.
+//
+// Since they won't have an ownerRef specified, reconciled secrets will not be deleted automatically on parent deletion.
+// To account for that, we add a label for best-effort garbage collection by the operator on parent resource deletion.
+func ReconcileSecretNoOwnerRef(c k8s.Client, expected corev1.Secret, theoreticalOwner metav1.Object) (corev1.Secret, error) {
+	// this function is similar to "ReconcileSecret", but:
+	// - we don't pass an owner
+	// - we remove the existing owner
+	// - we set additional labels to perform garbage collection on owner deletion (best-effort)
+	expected.Labels[SoftOwnerNamespaceLabel] = theoreticalOwner.GetNamespace()
+	expected.Labels[SoftOwnerNameLabel] = theoreticalOwner.GetName()
```

Comment:
Hm I think the problem is that the garbage collection part of the code only knows about the deleted resource name + namespace (as part of the `reconcile.Request`). We don't know anything about the parent ID since it doesn't exist anymore? (Maybe I'm misunderstanding what you mean).